### PR TITLE
Crop operator family to support per sample `crop` argument

### PR DESCRIPTION
**Pull request by jantonguirao**
## Category:

**New feature**

## Description:

- For relevant changes skip first commit (that was only moving to cc file)
- It enables support for per-sample `crop` argument
- Moves implementation to cc file

### Affected modules and functionalities:

Crop op family

### Key points relevant for the review:

CropAttr changes (last commit)

### Tests:

- [ ] Existing tests apply
- [ ] New tests added
  - [x] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A
test_crop.py::test_crop_arg_input

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

**Review by stiepan — APPROVED** [review]

File: dali/operators/image/crop/crop_attr.cc [hunk 1]
```diff
@@ -70,4 +69,159 @@ for ``crop_w``, ``crop_h``, and ``crop_d`` is incompatible with providing the fi
...
+void CropAttr::ProcessArguments(const OpSpec& spec, const ArgumentWorkspace* ws,
+                                std::size_t data_idx) {
+  int crop_arg_len = 0;
+  if (spec.HasTensorArgument("crop")) {
+    auto crop_arg = view<const float, 1>(ws->ArgumentInput("crop"))[data_idx];
+    crop_arg_len = crop_arg.shape[0];
+    DALI_ENFORCE(crop_arg_len >= 2 && crop_arg_len <= 3,
+                 "`crop` argument should have 2 or 3 elements depending on the input data shape");
+    int idx = 0;
+    has_crop_d_ = crop_arg_len == 3;
```
Comment (stiepan) [inline dali/operators/image/crop/crop_attr.cc:125]: Just a loose thought that at a first glance it seems slightly convoluted now. For instance where `has_crop_d_` comes from in particular tensor/spec case.

Comment (jantonguirao) [reply dali/operators/image/crop/crop_attr.cc:125]: You are right. I'll clean up

**Review by stiepan — APPROVED** [review]

**Comment by jantonguirao** [comment]
!build

File: dali/operators/image/crop/crop_attr.cc [hunk 2]
```diff
@@ -87,41 +85,47 @@ CropAttr::CropAttr(const OpSpec& spec) {
                  "`crop_d` argument must be provided together with `crop_w` and `crop_h`");
   }
 
-  size_t crop_arg_ndims = 0;
+  int crop_d = kNoCrop, crop_h = kNoCrop, crop_w = kNoCrop;
```
Comment (szalpal) [inline dali/operators/image/crop/crop_attr.cc:88]: `k` usually meant "constant", which these are `int`, not `const int`. Is that intentional?

File: dali/test/python/operator/test_crop.py [hunk 3]
```diff
@@ -568,3 +567,34 @@ def test_crop_empty_layout():
...
+        channel_dim = layout.find('C')
+        spatial_dims = [k for k in range(ndim) if k != channel_dim]
+        for i in range(len(out)):
+            expected = list(np.array(shape[i], dtype=np.int32))
+            actual = [np.array(out[i].shape())[k] for k in spatial_dims]
+            assert expected == actual, f"{expected} != {actual}"
+
+    for device in ['cpu', 'gpu']:
+        for layout in ['HWC', 'FHWC', 'CHW']:
+            yield check, device, layout
```
Comment (szalpal) [inline dali/test/python/operator/test_crop.py:600]: How about using `@params` and not abusing the test-generation as we did before? I believe the `@params` approach is way clearer, as you do not pass strage function as one of the params to `yield`...

Here's the example: https://github.com/NVIDIA/DALI/blob/0f6dba65674436ad350bf6621b6955492bebc530/dali/test/python/operator/test_remap.py#L106

File: dali/operators/image/crop/crop_attr.cc (hunk 2 above)
Comment (jantonguirao) [reply dali/operators/image/crop/crop_attr.cc:88]: it's just initialized to kNoCrop. Replaced later if applicable.
kNoCrop is constexpr:
```static constexpr int kNoCrop = -1;```

**Comment by jantonguirao** [comment]
!build

**Review by szalpal — APPROVED** [review]

**Pull request merged**